### PR TITLE
VDR task guide

### PR DESCRIPTION
**Pull request by merveenoyan**
Added VDR task guide as I noticed the model docs are too minimal for a solid use case

### CODE REVIEW

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+# Visual Document Retrieval
+
+Documents are essentially multimodal data, rich in charts, tables, visuals as well as text. Retrieving information from such documents is challenging as text retrieval models miss out on infographics, and image retrieval models lack granularity and document processing capabilities. Visual document retrieval is a new paradigm to solve information retrieval from documents. This comes in handy in all document retrieval use cases, a major one is multimodal retrieval augmented generation (RAG). These models essentially take in documents (as images) and texts and calculate similarity between each.
+
+This task guide aims to demonstrate how to index and retrieve documents using [ColPali](https://huggingface.co/vidore/colpali-v1.2-hf) with transformers.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:20]:
```suggestion
This guide demonstrates how to index and retrieve documents with [ColPali](../model_doc/colpali).  
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+Documents are essentially multimodal data, rich in charts, tables, visuals as well as text. Retrieving information from such documents is challenging as text retrieval models miss out on infographics, and image retrieval models lack granularity and document processing capabilities. Visual document retrieval is a new paradigm to solve information retrieval from documents. This comes in handy in all document retrieval use cases, a major one is multimodal retrieval augmented generation (RAG). These models essentially take in documents (as images) and texts and calculate similarity between each.
+
+This task guide aims to demonstrate how to index and retrieve documents using [ColPali](https://huggingface.co/vidore/colpali-v1.2-hf) with transformers.  
+
+<Tip>
+
+On a larger scale, you might want to index and retrieve your documents using a vector database. 
+
+</Tip>
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:26]:
```suggestion
> [!TIP]
> For large scale use cases, you may want to index and retrieve documents with a vector database.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+We will begin by installing transformers and datasets.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:28]:
```suggestion
Make sure Transformers and Datasets is installed.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+```bash
+pip install -q datasets transformers
+```
+
+We will index a dataset of documents related to UFO sightings. We filter the examples where our column of interest is missing. It contains several columns, we are interested in the column `specific_detail_query` where it contains short summary of the document, and `image` column that contains our documents.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:34]:
```suggestion
This dataset contains documents related to UFO sightings. Filter missing examples from the `specific_detail_query` column, which contains a short summary of the document and an `image` of the document.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+dataset
+```
+```
+Dataset({
+    features: ['image', 'raw_queries', 'broad_topical_query', 'broad_topical_explanation', 'specific_detail_query', 'specific_detail_explanation', 'visual_element_query', 'visual_element_explanation', 'parsed_into_json'],
+    num_rows: 2172
+})
+```
+
+Let's load the model and the tokenizer.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:51]:
```suggestion
Load the [`ColPaliProcessor`] and the [`ColPaliForRetrieval`] model.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+processor = ColPaliProcessor.from_pretrained(model_name)
+
+model = ColPaliForRetrieval.from_pretrained(
+    model_name,
+    torch_dtype=torch.bfloat16,
+    device_map="cuda",
+).eval()
+```
+
+We can simply infer with the model with text query like below. We can get the indexed embeddings with `embeddings` in our output. For image-to-text search, you can swap text with `images` argument and pass in images instead.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:68]:
```suggestion
Pass the text query to the processor and return the indexed text embeddings from the model. For image-to-text search, replace the `text` parameter in [`ColPaliProcessor`] with the `images` parameter to pass images.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+We can simply infer with the model with text query like below. We can get the indexed embeddings with `embeddings` in our output. For image-to-text search, you can swap text with `images` argument and pass in images instead. 
+
+```python
+inputs = processor(text="a document about Mars expedition").to("cuda")
+with torch.no_grad():
+  text_embeds = model(**inputs, return_tensors="pt").embeddings
+```
+
+We will do text-to-image retrieval. To do so, we need to index images offline, and during inference time, we get the text embeddings and simply get the closest image embedding to the query text embedding. Thus, we need to store image embeddings along with their images. A nice hack to index embeddings and write them on dataset is to use `dataset.map()` like below. We add an embeddings column that contains indexed embeddings. Since ColPali embeddings hold a lot of space, we need to remove them from GPU, so we detach them and write them as numpy vectors to store in CPU.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:76]:
```suggestion
Index the images offline, and during inference, return the query text embeddings to get its closest image embeddings.

Store the image and image embeddings by writing them to the dataset with [`~datasets.Dataset.map`] as shown below. Add an `embeddings` column that contains the indexed embeddings. ColPali embeddings take up a lot of storage, so remove them from the GPU and store them in the CPU as NumPy vectors.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+```python
+ds_with_embeddings = dataset.map(lambda example: {'embeddings': model(**processor(images=example["image"]).to("cuda"), return_tensors="pt").embeddings.to(torch.float32).detach().cpu().numpy()})
+```
+
+For the online inference part, we need a function to search image embeddings in batches and retrieve k-most-relevant images. Below function takes in the indexed dataset and text embedding, number of top results and batch size to get the best results and returns the indices in the dataset and their scores.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:82]:
```suggestion
For online inference, create a function to search the image embeddings in batches and retrieve the k-most relevant images. The function below returns the indices in the dataset and their scores for a given indexed dataset, text embeddings, number of top results, and the batch size.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    sorted_results = sorted(scores_and_indices, key=lambda x: -x[0])
+
+    topk = sorted_results[:k]
+    indices = [idx for _, idx in topk]
+    scores = [score for score, _ in topk]
+
+    return indices, scores
+```
+
+Let's infer. We should first get text embeddings and call our helper function above.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:110]:
```suggestion
Generate the text embeddings and pass them to the function above to return the dataset indices and scores.
```

File: docs/source/en/tasks/visual_document_retrieval.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+```
+
+```
+([440, 442, 443],
+ [14.370786666870117,
+  13.675487518310547,
+  12.9899320602417])
+```
+
+Now displaying the images we can see the model has returned Mars related documents.
```
Comment (stevhliu) [inline docs/source/en/tasks/visual_document_retrieval.md:126]:
```suggestion
Display the images to view the Mars related documents.
```

**Review by stevhliu — APPROVED** [review]
Super cool, thanks!

**Review by stevhliu — APPROVED** [review]
Thanks for the new task guide! 🤩
